File: HW5/task2.cpp
#include <iostream>
#include <string>

int SIZE = 1000;
int A = 7;
int PRIME = 1009;

struct Node {
	std::string key;
	std::string value;
	int empty = 1;
	Node *next = NULL;
};

class Map {
private:
	Node *arr = new Node[SIZE];

public:
	void put(std::string key, std::string value) {
		int i;
		Node *ptr;

		i = hash(key);
		ptr = &(this->arr[i]);
		while (ptr) {
			if (ptr->empty || ptr->key == key) {
				ptr->value = value;
				ptr->key = key;
				ptr->empty = 0;
				return ;
			}
			else {
				if (!ptr->next) {
					break;
				}
				ptr = ptr->next;
			}
		}
		Node *new_node = new Node;

		new_node->empty = 0;
		new_node->key = key;
		new_node->value = value;
		ptr->next = new_node;

	}


	void del(std::string key) {
		int i;
		Node *ptr;
		Node *next;
		Node *prev;
		Node *ss;

		i = hash(key);
		ss = prev = ptr = &(this->arr[i]);
		next = ptr->next;
		if (ptr->key == key) {
			if (ptr->next) {
				ptr = next;
				this->arr[i] = *ptr;
				return ;
			}
			else {
				this->arr[i].empty = 1;
				ptr = next;
			}
		}
		while (ptr) {
			if (ptr->key == key) {
				prev->next = next;
				delete ptr;
				return ;
			}
			prev = ptr;
			ptr = next;
			if (ptr) {
				next = ptr->next;
			}
		}
	}

	std::string get(std::string key) {
		int i;
		Node *ptr;

		i = hash(key);
		ptr = &(this->arr[i]);
		while (ptr) {
			if (ptr->key == key && ptr->empty == 0) {
				return ptr->value;
			}
			ptr = ptr->next;
		}
		return "none";
	}

	int hash(std::string key) {
		int res;

		res = 0;
		for (int i = 0; i < key.length(); i++) {
			res = (res * A + (key[i] - '0')) % PRIME;
		}
		res = res % SIZE;
		return (res);
	}
};


int main() {
	Map my_map;

	std::string command, key, value;
	while (std::cin >> command) {
		if (command == "put") {
			std::cin >> key >> value;
			my_map.put(key, value);
		}
		else if (command == "get") {
			std::cin >> key;
			std::cout << my_map.get(key) << "\n";
		}
		else if (command == "delete") {
			std::cin >> key;
			my_map.del(key);
		}
	}
	return (0);
}

File: HW5/task2.py
import sys

SIZE = 1000
A = 7
PRIME = 1009

class Node(object):

    def __init__(self, key=None, data=None, next_node=None):
        self.data = data
        self.key = key
        self.next_node = next_node

    def get_data(self):
        return self.data

    def get_next(self):
        return self.next_node

    def set_next(self, new_next):
        self.next_node = new_next

class LinkedList(object):
    def __init__(self, head=None):
        self.head = head

    def insert(self, data):
        new_node = Node(data)
        new_node.set_next(self.head)
        self.head = new_node

    def size(self):
        current = self.head
        count = 0
        while current:
            count += 1
            current = current.get_next()
        return count

    def search(self, data):
        current = self.head
        found = False
        while current and found is False:
            if current.get_data() == data:
                found = True
            else:
                current = current.get_next()
        if current is None:
            None
        return current

    def delete(self, data):
        current = self.head
        previous = None
        found = False
        while current and found is False:
            if current.get_data() == data:
                found = True
            else:
                previous = current
                current = current.get_next()
        if current is None:
            return None
        if previous is None:
            self.head = current.get_next()
        else:
            previous.set_next(current.get_next())


class HashMap():
    def __init__(self):
        self.arr = [None for i in range(SIZE)]
        self.size = SIZE

    def put(self, key, value):
        i = self.hash(value)
        if not self.arr[i]:
            self.arr[i] = LinkedList()
        self.arr[i].insert(value)

    def delete(self, value):
        if not self.arr[i].delete(value):
            self.arr[i] = None

    def get(self,value):
        if self.arr[i].get(value):
            self.arr[i].get(value).get_data()


    def hash(self, key):
        res = 0
        for i in str(key):
            res = (res * A + ord(i)) % PRIME
        res = res % self.size
        return int(res)
    

if __name__ == "__main__":
    my_map = HashMap()
    ans = []
    for line in sys.stdin.read().splitlines():
        if line.split()[0] == "put":
            my_map.put(line.split()[1])
        elif line.split()[0] == "get":
            if my_map.get(line.split()[1])[0]:
                ans.append("true")
            else:
                ans.append("false")
        elif line.split()[0] == "delete":
            my_map.delete(line.split()[1])
    sys.stdout.buffer.write("\n".join(ans).encode())
